[BB-1506] Use a callable as the default value for CourseEnrollment.mode

lgp171188 opened the pull request:
The previous behaviour of using a variable causes Django to complain
about changes not reflected in a migration when the default course
mode slug is changed.

**JIRA tickets**: If there is a related publicly viewable JIRA ticket, mention that here. Also mention the OSPR ticket
number here once it gets created (after you open the PR), if applicable. Otherwise omit this.

**Dependencies**: None

**Sandbox URL**: TBD - sandbox is being provisioned.

**Merge deadline**: None

**Testing instructions**:
1. In the `master` devstack, set the `COURSE_MODE_DEFAULTS` variable in the LMS configuration file (`/edx/etc/lms.yml`) with values like below.

    ```
    COURSE_MODE_DEFAULTS:
      bulk_sku: null
      currency: 'usd'
      description: null
      expiration_datetime: null
      min_price: 0
      name: 'Honor'
      sku: None
      slug: 'honor'
      suggested_prices: ''
    ```
2. Restart LMS.
3. Run `python manage.py lms makemigrations student`. A new migration will be created for the `CourseEnrollment` model setting the default value for the `mode` field to `"honor"`.
4. Repeat steps 1 to 3 with the source branch of this PR. There should be no additional migrations created with value of slug set to `'audit'` (the default), `'honor'` or any other mode slug.

**Reviewers**
- [ ] @giovannicimolin 
- [ ] edX reviewer[s] TBD

**Settings**
```yaml
EDXAPP_ENV_EXTRA:
  COURSE_MODE_DEFAULTS:
    bulk_sku: null
    currency: 'usd'
    description: null
    expiration_datetime: null
    min_price: 0
    name: 'Honor'
    sku: None
    slug: 'honor'
    suggested_prices: ''
```

edx-webhook commented:
Thanks for the pull request, @lgp171188! I've created [OSPR-3754](https://openedx.atlassian.net/browse/OSPR-3754) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

natabene commented:
@lgp171188 Thank you for your contribution. Please let me know once all checks are green and this is ready for our review.

lgp171188 commented:
> Your PR has finished running tests. There were no failures.

@natabene, the bot has mentioned that there were no failures, however the `Datree Smart Policy` check has failed and the details page has no information available at all to be able to check and fix issues, if any.

natabene commented:
@lgp171188 I saw similar issue elsewhere, maybe there is a problem not related to your code.

natabene commented:
@ormsbee Can you give this a look?

giovannicimolin commented:
@ormsbee I ran `python manage.py lms sqlmigrate student 0023
` on **lms-shell** and got this as output:
```
BEGIN;
--
-- Alter field mode on courseenrollment
--
ALTER TABLE `student_courseenrollment` ALTER COLUMN `mode` SET DEFAULT 'honor';
ALTER TABLE `student_courseenrollment` ALTER COLUMN `mode` DROP DEFAULT;
--
-- Alter field mode on historicalcourseenrollment
--
ALTER TABLE `student_courseenrollment_history` ALTER COLUMN `mode` SET DEFAULT 'honor';
ALTER TABLE `student_courseenrollment_history` ALTER COLUMN `mode` DROP DEFAULT;
COMMIT;
```
I think it just drops the default value from the schema and lets Django handle it using the callable.

lgp171188 commented:
@ormsbee, did you get a chance to check @giovannicimolin's response to your question?

lgp171188 commented:
@ormsbee, ping! Did you get a chance to review this?

natabene commented:
@lgp171188 Sorry, we have limited bandwidth now, so won't be able to review soon.

lgp171188 commented:
@ormsbee, @natabene, any updates?

natabene commented:
@lgp171188 No, no updates yet.

giovannicimolin commented:
jenkins run all

giovannicimolin commented:
@lgp171188 Rerunning test suite since there are new test suites.

giovannicimolin commented:
@lgp171188 Can you rebase your PR and re-run the tests when you're back?

lgp171188 commented:
@giovannicimolin, I have rebased the PR on top of the latest changes in the master branch, fixed the migration number as there have been newer migrations created after this PR was created. The CI checks have passed now. So this is ready for prioritization.

natabene commented:
@lgp171188 thanks, I will queue this for review.

edx-webhook commented:
@lgp171188 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/21166&entry.1484655318=https://github.com/lgp171188&entry.752974735=2019-07-19+13:36&entry.1917517419=2019-12-26+15:25&entry.2133058324=Yes) so we can improve your experience in the future.
